File: epi_judge_cpp/task_pairing.cc
#include <tuple>
#include <vector>

#include "test_framework/fmt_print.h"
#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
using std::vector;

struct PairedTasks {
  int task_1, task_2;
};

vector<PairedTasks> OptimumTaskAssignment(vector<int> task_durations) {
  // TODO - you fill in here.
  return {};
}

namespace test_framework {
template <>
struct SerializationTrait<PairedTasks> : UserSerTrait<PairedTasks, int, int> {};
}  // namespace test_framework

bool operator==(const PairedTasks& lhs, const PairedTasks& rhs) {
  return std::tie(lhs.task_1, lhs.task_2) == std::tie(rhs.task_1, rhs.task_2);
}

std::ostream& operator<<(std::ostream& out, const PairedTasks& t) {
  return PrintTo(out, std::make_tuple(t.task_1, t.task_2));
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"task_durations"};
  return GenericTestMain(args, "task_pairing.cc", "task_pairing.tsv",
                         &OptimumTaskAssignment, DefaultComparator{},
                         param_names);
}

File: epi_judge_java/epi/TaskPairing.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.EpiUserType;
import epi.test_framework.GenericTest;
import epi.test_framework.EpiTestComparator;

import java.util.List;
import java.util.Comparator;

public class TaskPairing {
  @EpiUserType(ctorParams = {Integer.class, Integer.class})

  public static class PairedTasks {
    public Integer task1;
    public Integer task2;

    public PairedTasks(Integer task1, Integer task2) {
      this.task1 = task1;
      this.task2 = task2;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }

      PairedTasks that = (PairedTasks)o;

      return task1.equals(that.task1) && task2.equals(that.task2) ||
        task1.equals(that.task2) && task2.equals(that.task1);
    }

    @Override
    public String toString() {
      return "[" + task1 + ", " + task2 + "]";
    }
  }

  private static class PairedTasksComparator implements Comparator<PairedTasks> {
    @Override
    public int compare(PairedTasks p1, PairedTasks p2) {
      if (p1 == null) {
        return p2 == null ? 0 : -1;
      }

      if (p1.equals(p2)) {
        return 0;
      }

      Integer minP1 = Math.min(p1.task1, p1.task2);
      Integer minP2 = Math.min(p2.task1, p2.task2);
      int minCompare = minP1.compareTo(minP2);

      if (minCompare != 0) {
        return minCompare;
      } else {
        Integer maxP1 = Math.max(p1.task1, p1.task2);
        Integer maxP2 = Math.max(p2.task1, p2.task2);
        return maxP1.compareTo(maxP2);
      }
    }
  }

  @EpiTest(testDataFile = "task_pairing.tsv")

  public static List<PairedTasks>
  optimumTaskAssignment(List<Integer> taskDurations) {
    // TODO - you fill in here.
    return null;
  }

  @EpiTestComparator
  public static boolean comp(List<PairedTasks> expected, List<PairedTasks> result) {
    if (result == null) {
      return false;
    }
    result.sort(new PairedTasksComparator());
    expected.sort(new PairedTasksComparator());
    return expected.equals(result);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "TaskPairing.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
